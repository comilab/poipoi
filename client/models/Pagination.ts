export default class Pagination {
  currentPage: number
  from: number
  lastPage: number
  path: string
  perPage: number
  to: number
  total: number

  constructor (data: any) {
    this.currentPage = data.currentPage
    this.from = data.from
    this.lastPage = data.lastPage
    this.path = data.path
    this.perPage = data.perPage
    this.to = data.to
    this.total = data.total
  }
}
