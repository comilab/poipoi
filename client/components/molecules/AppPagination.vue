<template lang="pug">
.flex
  AppButton.mx-2(
    color="white",
    :icon="faAngleDoubleLeft",
    @click="onClick(1)",
    :class="{ 'pointer-events-none': pagination.currentPage === 1 }",
    data-testid="page-first"
  )
  .mx-1(
    v-if="range[0] > 1",
    data-testid="left-ellipsis"
  ) …
  AppButton.mx-2(
    v-for="page in range",
    :key="page",
    :color="getColor(page)",
    square,
    @click="onClick(page)",
    :data-testid="`page-${page}`"
  ) {{ page }}
  .mx-1(
    v-if="range.slice(-1).pop() < pagination.lastPage",
    data-testid="right-ellipsis"
  ) …
  AppButton.mx-2(
    color="white",
    :icon="faAngleDoubleRight",
    @click="onClick(pagination.lastPage)",
    :class="{ 'pointer-events-none': pagination.currentPage === pagination.lastPage }",
    data-testid="page-last"
  )
</template>

<script lang="ts">
import { defineComponent, computed } from 'nuxt-composition-api'
import lodashRange from 'lodash/range'
import { faAngleDoubleLeft, faAngleDoubleRight } from '@fortawesome/free-solid-svg-icons'

import Pagination from '~/models/Pagination'
import AppButton from '~/components/atoms/AppButton.vue'

export default defineComponent({
  components: {
    AppButton
  },
  props: {
    pagination: {
      type: Pagination,
      required: true
    }
  },
  setup (props, { emit }) {
    const range = computed(() => {
      const from = Math.max(
        1,
        Math.min(
          props.pagination.currentPage - 2,
          props.pagination.lastPage - 4
        )
      )

      return lodashRange(
        from,
        Math.min(from + 5, props.pagination.lastPage + 1)
      )
    })

    const getColor = (page: number) => {
      return page === props.pagination.currentPage ? 'blue' : 'white'
    }

    const onClick = (page: number) => {
      emit('selected', page)
    }

    return {
      range,
      getColor,
      onClick,
      faAngleDoubleLeft,
      faAngleDoubleRight
    }
  }
})
</script>
