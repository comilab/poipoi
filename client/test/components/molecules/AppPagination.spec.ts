import { fireEvent, render } from '@testing-library/vue'

import AppPagination from '~/components/molecules/AppPagination.vue'
import Pagination from '~/models/Pagination'
import paginationFactory from '~/test/factories/pagination'

describe('components/molecules/AppPagination', () => {
  let options: any

  beforeEach(() => {
    options = {
      props: {
        pagination: paginationFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true
      }
    }
  })

  test('renders correctly', () => {
    options.props.pagination = new Pagination({
      currentPage: 5,
      from: 51,
      lastPage: 10,
      path: '/',
      perPage: 10,
      to: 60,
      total: 100
    })
    const { container } = render(AppPagination, options)

    expect(container).toMatchSnapshot()
  })

  describe('props', () => {
    describe('pagination', () => {
      test('currentPageが1ならpage-firstはクリック不可', () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 1
        })
        const { getByTestId } = render(AppPagination, options)

        expect(getByTestId('page-first')).toHaveClass('pointer-events-none')
      })

      test('currentPageとlastPageが等しければpage-lastはクリック不可', () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 1,
          lastPage: 1
        })
        const { getByTestId } = render(AppPagination, options)

        expect(getByTestId('page-last')).toHaveClass('pointer-events-none')
      })

      test('ページボタンはcurrentPageを中心に5個まで表示', () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 7,
          lastPage: 100
        })
        const { queryByTestId } = render(AppPagination, options)
        expect(queryByTestId('page-4')).not.toBeInTheDocument()
        expect(queryByTestId('page-5')).toBeInTheDocument()
        expect(queryByTestId('page-6')).toBeInTheDocument()
        expect(queryByTestId('page-7')).toBeInTheDocument()
        expect(queryByTestId('page-8')).toBeInTheDocument()
        expect(queryByTestId('page-9')).toBeInTheDocument()
        expect(queryByTestId('page-10')).not.toBeInTheDocument()
      })

      test('currentPageが2以下の場合は1ページ目から5ページ目まで表示', () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 2,
          lastPage: 100
        })
        const { queryByTestId } = render(AppPagination, options)
        expect(queryByTestId('page-1')).toBeInTheDocument()
        expect(queryByTestId('page-2')).toBeInTheDocument()
        expect(queryByTestId('page-3')).toBeInTheDocument()
        expect(queryByTestId('page-4')).toBeInTheDocument()
        expect(queryByTestId('page-5')).toBeInTheDocument()
        expect(queryByTestId('page-6')).not.toBeInTheDocument()
      })

      test('lastPageが4以下の場合は1ページ目からlastPageページ目まで表示', () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 4,
          lastPage: 4
        })
        const { queryByTestId } = render(AppPagination, options)
        expect(queryByTestId('page-1')).toBeInTheDocument()
        expect(queryByTestId('page-2')).toBeInTheDocument()
        expect(queryByTestId('page-3')).toBeInTheDocument()
        expect(queryByTestId('page-4')).toBeInTheDocument()
        expect(queryByTestId('page-5')).not.toBeInTheDocument()
      })

      test('lastPageを超えて描画しない', () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 1,
          lastPage: 3
        })
        const { queryByTestId } = render(AppPagination, options)
        expect(queryByTestId('page-4')).not.toBeInTheDocument()
      })

      test('ページボタンが2以上から始まっている場合はleft-ellipsisを表示', async () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 1
        })
        const { queryByTestId, updateProps } = render(AppPagination, options)
        expect(queryByTestId('left-ellipsis')).not.toBeInTheDocument()

        await updateProps({
          pagination: paginationFactory.build({
            currentPage: 4,
            lastPage: 6
          })
        })
        expect(queryByTestId('page-1')).not.toBeInTheDocument()
        expect(queryByTestId('left-ellipsis')).toBeInTheDocument()
      })

      test('最後に表示されるページボタンがlastPageと異なる場合はright-ellipsisを表示', async () => {
        options.props.pagination = paginationFactory.build({
          currentPage: 5,
          lastPage: 7
        })
        const { queryByTestId, updateProps } = render(AppPagination, options)
        expect(queryByTestId('right-ellipsis')).not.toBeInTheDocument()

        await updateProps({
          pagination: paginationFactory.build({
            currentPage: 5,
            lastPage: 8
          })
        })
        expect(queryByTestId('right-ellipsis')).toBeInTheDocument()
      })
    })
  })

  describe('event', () => {
    test('page-firstをクリックするとselected(1)イベントを発火', async () => {
      options.props.pagination = paginationFactory.build({
        currentPage: 2
      })
      const { getByTestId, emitted } = render(AppPagination, options)
      await fireEvent.click(getByTestId('page-first'))

      expect(emitted().selected).toHaveLength(1)
      expect(emitted().selected[0]).toStrictEqual([1])
    })

    test('page-lastをクリックするとselected(lastPage)イベントを発火', async () => {
      options.props.pagination = paginationFactory.build({
        currentPage: 1,
        lastPage: 10
      })
      const { getByTestId, emitted } = render(AppPagination, options)
      await fireEvent.click(getByTestId('page-last'))

      expect(emitted().selected).toHaveLength(1)
      expect(emitted().selected[0]).toStrictEqual([10])
    })

    test('page-{num}をクリックするとselected(num)イベントを発火', async () => {
      options.props.pagination = paginationFactory.build({
        currentPage: 1,
        lastPage: 10
      })
      const { getByTestId, emitted } = render(AppPagination, options)
      await fireEvent.click(getByTestId('page-2'))

      expect(emitted().selected).toHaveLength(1)
      expect(emitted().selected[0]).toStrictEqual([2])
    })
  })
})
