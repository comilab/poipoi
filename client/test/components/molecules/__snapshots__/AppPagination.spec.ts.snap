// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppPagination renders correctly 1`] = `
<div>
  <div
    class="flex"
  >
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none mx-2 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded"
      data-testid="page-first"
      type="button"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      />
    </button>
    <div
      class="mx-1"
      data-testid="left-ellipsis"
    >
      …
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none mx-2 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded"
      data-testid="page-3"
      type="button"
    >
      <!---->
      <div
        class=""
        data-testid="text"
      >
        3
      </div>
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none mx-2 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded"
      data-testid="page-4"
      type="button"
    >
      <!---->
      <div
        class=""
        data-testid="text"
      >
        4
      </div>
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none mx-2 bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 square-md rounded"
      data-testid="page-5"
      type="button"
    >
      <!---->
      <div
        class=""
        data-testid="text"
      >
        5
      </div>
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none mx-2 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded"
      data-testid="page-6"
      type="button"
    >
      <!---->
      <div
        class=""
        data-testid="text"
      >
        6
      </div>
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none mx-2 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded"
      data-testid="page-7"
      type="button"
    >
      <!---->
      <div
        class=""
        data-testid="text"
      >
        7
      </div>
    </button>
    <div
      class="mx-1"
      data-testid="right-ellipsis"
    >
      …
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none mx-2 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded"
      data-testid="page-last"
      type="button"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      />
    </button>
  </div>
</div>
`;
